### PR TITLE
[dlive] add extractor

### PR DESCRIPTION
**Pull request by ealgase**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This pull request supports dlive VOD's.  It doesn't yet support dlive live streams.

### CODE REVIEW

**Comment by ealgase** [comment]
I've also added support for livestreams.  I have a feeling that while everything I'm doing *works*, I'm probably doing some stuff wrong, so feedback would be appreciated!

File: youtube_dl/extractor/dlive.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+import re
+import six
```
Comment (dstftw) [inline youtube_dl/extractor/dlive.py]: No.

File: youtube_dl/extractor/dlive.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        'url': 'https://dlive.tv/p/pdp+J0H49ZgZR',
+        'md5': '43c96ce082b3fd2e1dae07fa73f137d8',
+        'info_dict': {
+            'id': 'J0H49ZgZR',
+            'ext': 'mp4',
+            'title': '#SupportCreators #StreamOnDLive',
+            'description': 'Replay created by pdp | DLive',
+            'uploader': 'pdp',
+            'uploader_id': 'pdp',
+            # TODO more properties, either as:
```
Comment (dstftw) [inline youtube_dl/extractor/dlive.py]: Do.

File: youtube_dl/extractor/dlive.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        description = self._html_search_meta('og:description', webpage, default=None)
+        video_apollo_state_json = self._search_regex(r'window\.__APOLLO_STATE__=(.+?);', webpage, vod_id)
+        video_parsed_json = self._parse_json(video_apollo_state_json, vod_id)
+        video_m3u8_url_1 = video_parsed_json['defaultClient']
+        # id of the JSON element varies so using the six package is the most future proof and consistent way of doing this
+        extracted_json_info = six.next(six.itervalues(video_m3u8_url_1))
+        m3u8_url = extracted_json_info['playbackUrl']
+        title = extracted_json_info['title']
+        parsed_formats = self._extract_m3u8_formats(m3u8_url, vod_id, ext='mp4')
+        # TODO more code goes here, for example ...
```
Comment (dstftw) [inline youtube_dl/extractor/dlive.py]: Do.

File: youtube_dl/extractor/dlive.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        return {
+            'id': vod_id,
+            'title': title,
+            'uploader': uploader_id,
+            'uploader_id': uploader_id,
+            'formats': parsed_formats,
+            'description': description,
+            'thumbnail': thumbnail,
+            # TODO more properties (see youtube_dl/extractor/common.py)
```
Comment (dstftw) [inline youtube_dl/extractor/dlive.py]: Do.

File: youtube_dl/extractor/dlive.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        mobj = re.match(self._VALID_URL, url)
+        channel = mobj.group('channel')
+        webpage = self._download_webpage(url, channel)
+        thumbnail = self._html_search_meta('og:image', webpage, default=None)
+        video_apollo_state_json = self._search_regex(r'window\.__APOLLO_STATE__=(.+?);', webpage, channel)
+        video_parsed_json = self._parse_json(video_apollo_state_json, channel)
+        video_info = video_parsed_json['defaultClient']
+        # id of the JSON element varies so using the six package is the most future proof and consistent way of doing this
+        extracted_json_info = six.next(six.itervalues(video_info))
+        uploader_id = extracted_json_info['username']
```
Comment (dstftw) [inline youtube_dl/extractor/dlive.py]: Read coding conventions on optional metadata.

**Comment by ealgase** [comment]
I just realized that this isn't compatible with python2.  While I've fixed all of the other issues (and can push them), I think that since python2 fundamentally loads JSON differently, making this extractor work with it while following youtube-dl's guidelines is impossible.

The original JSON data (to be loaded)
```json
{"defaultClient":{"user:pdp":{"id":"user:pdp","avatar":"https:\u002F\u002Fimages.prd.dlivecdn.com\u002Favatar\u002F6c1fccb3-5a6c-11e9-97a3-12567f05983b","__typename":"User","displayname":"PewDiePie","partnerStatus":"VERIFIED_PARTNER","username":"pdp","followers":{"type":"id","generated":true,"id":"$user:pdp.followers","typename":"UserConnection"},"canSubscribe":true,"banStatus":"NO_BAN","about":"I am PewDiePie. This is my official channel on DLive. \n\nJoin #SupportCreators campaign with me now at https:\u002F\u002Fdlive.gg\u002Fpewdiepie\n\nI will start live streaming exclusively on DLive on April 14th. In my first stream, I will support creators by donating up to $50,000 to a max of 100 creators. Follow any creator on DLive to increase the donation amount!\n\n","livestream":{"type":"id","generated":false,"id":"livestream:pdp+o5zYlcgZR","typename":"Livestream"},"hostingLivestream":null,"videos":{"type":"id","generated":true,"id":"$user:pdp.videos","typename":"VideoConnection"},"pastBroadcasts":{"type":"id","generated":true,"id":"$user:pdp.pastBroadcasts","typename":"PastBroadcastConnection"},"following":{"type":"id","generated":true,"id":"$user:pdp.following","typename":"UserConnection"}},"$user:pdp.followers":{"totalCount":142067,"__typename":"UserConnection"},"livestream:pdp+o5zYlcgZR":{"permlink":"pdp+o5zYlcgZR","ageRestriction":false,"id":"livestream:pdp+o5zYlcgZR","creator":{"type":"id","generated":true,"id":"$livestream:pdp+o5zYlcgZR.creator","typename":"User"},"title":"#SupportCreators #StreamOnDLive","totalReward":"4870986200","watchingCount":2664,"thumbnailUrl":"https:\u002F\u002Fimages.prd.dlivecdn.com\u002Flive-thumbnail\u002Fd58aa1b7-5e20-11e9-bcb8-f2baabbfad97","lastUpdatedAt":"1555183121000","__typename":"Livestream","watchTime({\"add\":false})":true,"category":{"type":"id","generated":false,"id":"category:12646","typename":"Category"},"content":"","disableAlert":false,"language":{"type":"id","generated":true,"id":"$livestream:pdp+o5zYlcgZR.language","typename":"Language"}},"$livestream:pdp+o5zYlcgZR.creator":{"username":"pdp","displayname":"PewDiePie","avatar":"https:\u002F\u002Fimages.prd.dlivecdn.com\u002Favatar\u002F6c1fccb3-5a6c-11e9-97a3-12567f05983b","__typename":"User","partnerStatus":"VERIFIED_PARTNER"},"category:12646":{"id":"category:12646","backendID":12646,"title":"DLive","imgUrl":"https:\u002F\u002Fimages.prd.dlivecdn.com\u002Fcategory\u002Ff938qp9jfaksjdfy9","watchingCount":4644,"__typename":"Category"},"$livestream:pdp+o5zYlcgZR.language":{"language":"English","__typename":"Language"},"$user:pdp.videos":{"totalCount":0,"__typename":"VideoConnection"},"$user:pdp.pastBroadcasts":{"totalCount":5,"__typename":"PastBroadcastConnection"},"$user:pdp.following":{"totalCount":0,"__typename":"UserConnection"},"ROOT_QUERY":{"userByDisplayName({\"displayname\":\"PewDiePie\"})":{"type":"id","generated":false,"id":"user:pdp","typename":"User"}}}}
```
The way it's loaded in python3 (correct)
```python
{'defaultClient': {'user:pdp': {'id': 'user:pdp', 'avatar': 'https://images.prd.dlivecdn.com/avatar/6c1fccb3-5a6c-11e9-97a3-12567f05983b', '__typename': 'User', 'displayname': 'PewDiePie', 'partnerStatus': 'VERIFIED_PARTNER', 'username': 'pdp', 'followers': {'type': 'id', 'generated': True, 'id': '$user:pdp.followers', 'typename': 'UserConnection'}, 'canSubscribe': True, 'banStatus': 'NO_BAN', 'about': 'I am PewDiePie. This is my official channel on DLive. \n\nJoin #SupportCreators campaign with me now at https://dlive.gg/pewdiepie\n\nI will start live streaming exclusively on DLive on April 14th. In my first stream, I will support creators by donating up to $50,000 to a max of 100 creators. Follow any creator on DLive to increase the donation amount!\n\n', 'livestream': {'type': 'id', 'generated': False, 'id': 'livestream:pdp+o5zYlcgZR', 'typename': 'Livestream'}, 'hostingLivestream': None, 'videos': {'type': 'id', 'generated': True, 'id': '$user:pdp.videos', 'typename': 'VideoConnection'}, 'pastBroadcasts': {'type': 'id', 'generated': True, 'id': '$user:pdp.pastBroadcasts', 'typename': 'PastBroadcastConnection'}, 'following': {'type': 'id', 'generated': True, 'id': '$user:pdp.following', 'typename': 'UserConnection'}}, '$user:pdp.followers': {'totalCount': 142067, '__typename': 'UserConnection'}, 'livestream:pdp+o5zYlcgZR': {'permlink': 'pdp+o5zYlcgZR', 'ageRestriction': False, 'id': 'livestream:pdp+o5zYlcgZR', 'creator': {'type': 'id', 'generated': True, 'id': '$livestream:pdp+o5zYlcgZR.creator', 'typename': 'User'}, 'title': '#SupportCreators #StreamOnDLive', 'totalReward': '4870986200', 'watchingCount': 2664, 'thumbnailUrl': 'https://images.prd.dlivecdn.com/live-thumbnail/d58aa1b7-5e20-11e9-bcb8-f2baabbfad97', 'lastUpdatedAt': '1555183121000', '__typename': 'Livestream', 'watchTime({"add":false})': True, 'category': {'type': 'id', 'generated': False, 'id': 'category:12646', 'typename': 'Category'}, 'content': '', 'disableAlert': False, 'language': {'type': 'id', 'generated': True, 'id': '$livestream:pdp+o5zYlcgZR.language', 'typename': 'Language'}}, '$livestream:pdp+o5zYlcgZR.creator': {'username': 'pdp', 'displayname': 'PewDiePie', 'avatar': 'https://images.prd.dlivecdn.com/avatar/6c1fccb3-5a6c-11e9-97a3-12567f05983b', '__typename': 'User', 'partnerStatus': 'VERIFIED_PARTNER'}, 'category:12646': {'id': 'category:12646', 'backendID': 12646, 'title': 'DLive', 'imgUrl': 'https://images.prd.dlivecdn.com/category/f938qp9jfaksjdfy9', 'watchingCount': 4644, '__typename': 'Category'}, '$livestream:pdp+o5zYlcgZR.language': {'language': 'English', '__typename': 'Language'}, '$user:pdp.videos': {'totalCount': 0, '__typename': 'VideoConnection'}, '$user:pdp.pastBroadcasts': {'totalCount': 5, '__typename': 'PastBroadcastConnection'}, '$user:pdp.following': {'totalCount': 0, '__typename': 'UserConnection'}, 'ROOT_QUERY': {'userByDisplayName({"displayname":"PewDiePie"})': {'type': 'id', 'generated': False, 'id': 'user:pdp', 'typename': 'User'}}}}
```
The way it's loaded in python2 (incorrect)
```python
{u'defaultClient': {u'ROOT_QUERY': {u'userByDisplayName({"displayname":"PewDiePie"})': {u'typename': u'User', u'generated': False, u'type': u'id', u'id': u'user:pdp'}}, u'$livestream:pdp+o5zYlcgZR.language': {u'__typename': u'Language', u'language': u'English'}, u'$user:pdp.videos': {u'totalCount': 0, u'__typename': u'VideoConnection'}, u'$user:pdp.pastBroadcasts': {u'totalCount': 5, u'__typename': u'PastBroadcastConnection'}, u'$user:pdp.following': {u'totalCount': 0, u'__typename': u'UserConnection'}, u'user:pdp': {u'username': u'pdp', u'about': u'I am PewDiePie. This is my official channel on DLive. \n\nJoin #SupportCreators campaign with me now at https://dlive.gg/pewdiepie\n\nI will start live streaming exclusively on DLive on April 14th. In my first stream, I will support creators by donating up to $50,000 to a max of 100 creators. Follow any creator on DLive to increase the donation amount!\n\n', u'displayname': u'PewDiePie', u'videos': {u'typename': u'VideoConnection', u'generated': True, u'type': u'id', u'id': u'$user:pdp.videos'}, u'canSubscribe': True, u'banStatus': u'NO_BAN', u'pastBroadcasts': {u'typename': u'PastBroadcastConnection', u'generated': True, u'type': u'id', u'id': u'$user:pdp.pastBroadcasts'}, u'__typename': u'User', u'hostingLivestream': None, u'partnerStatus': u'VERIFIED_PARTNER', u'followers': {u'typename': u'UserConnection', u'generated': True, u'type': u'id', u'id': u'$user:pdp.followers'}, u'avatar': u'https://images.prd.dlivecdn.com/avatar/6c1fccb3-5a6c-11e9-97a3-12567f05983b', u'following': {u'typename': u'UserConnection', u'generated': True, u'type': u'id', u'id': u'$user:pdp.following'}, u'livestream': {u'typename': u'Livestream', u'generated': False, u'type': u'id', u'id': u'livestream:pdp+o5zYlcgZR'}, u'id': u'user:pdp'}, u'$livestream:pdp+o5zYlcgZR.creator': {u'username': u'pdp', u'partnerStatus': u'VERIFIED_PARTNER', u'displayname': u'PewDiePie', u'avatar': u'https://images.prd.dlivecdn.com/avatar/6c1fccb3-5a6c-11e9-97a3-12567f05983b', u'__typename': u'User'}, u'category:12646': {u'title': u'DLive', u'__typename': u'Category', u'backendID': 12646, u'watchingCount': 4644, u'id': u'category:12646', u'imgUrl': u'https://images.prd.dlivecdn.com/category/f938qp9jfaksjdfy9'}, u'livestream:pdp+o5zYlcgZR': {u'category': {u'typename': u'Category', u'generated': False, u'type': u'id', u'id': u'category:12646'}, u'language': {u'typename': u'Language', u'generated': True, u'type': u'id', u'id': u'$livestream:pdp+o5zYlcgZR.language'}, u'title': u'#SupportCreators #StreamOnDLive', u'ageRestriction': False, u'__typename': u'Livestream', u'content': u'', u'disableAlert': False, u'creator': {u'typename': u'User', u'generated': True, u'type': u'id', u'id': u'$livestream:pdp+o5zYlcgZR.creator'}, u'lastUpdatedAt': u'1555183121000', u'watchTime({"add":false})': True, u'watchingCount': 2664, u'permlink': u'pdp+o5zYlcgZR', u'thumbnailUrl': u'https://images.prd.dlivecdn.com/live-thumbnail/d58aa1b7-5e20-11e9-bcb8-f2baabbfad97', u'totalReward': u'4870986200', u'id': u'livestream:pdp+o5zYlcgZR'}, u'$user:pdp.followers': {u'totalCount': 142067, u'__typename': u'UserConnection'}}}
```

**Comment by dstftw** [comment]
Wrong. Both are the same JSON. Internal key ordering in dicts does not matter anything.

**Pull request closed without merging**